chef/automate
# Check-in time series total nodes

### :nut_and_bolt: Description: What code changed, and why?
Adding the total number of nodes for each daily range. To calculate the number of nodes for each range find the number of nodes that have been deleted and created before the end of the range. Subtract the number of nodes created from the number of nodes delete and you have the total number of nodes. To find the number of nodes created, a create date was added to the node-state. During ingestion, if the node is new it adds the create date of the time now.  If the node already exists it does not update the create date. 

```
{
  "counts": [
    {
      "start": "2020-03-20T19:00:00Z",
      "end": "2020-03-21T18:59:59Z",
      "count": 0,
      "total": 0 <-------
    },
    {
      "start": "2020-03-21T19:00:00Z",
      "end": "2020-03-22T18:59:59Z",
      "count": 5,
      "total": 5 <-------
    },
    {
      "start": "2020-03-22T19:00:00Z",
      "end": "2020-03-23T18:59:59Z",
      "count": 6,
      "total": 6  <-------
    }
  ]
}
```

Note: Copied one of olivere's go files to add the additional "missing" option to allow for counting nodes that do not have a create date. https://github.com/olivere/elastic/blob/release-branch.v6/search_aggs_bucket_date_range.go

Below is a link the the "missing" option information.
https://www.elastic.co/guide/en/elasticsearch/reference/7.0/search-aggregations-bucket-daterange-aggregation.html#_missing_values_3

### :chains: Related Resources

past PR - https://github.com/chef/automate/pull/3137
issue - https://github.com/chef/automate/issues/2995

### :+1: Definition of Done
The correct total of nodes is added to the check-in counts time-series requests. 

### :athletic_shoe: How to Build and Test the Change
1. `start_all_services`
1. Add some nodes before the PR with `chef_load_nodes 5`
1. Run this command to get the time series. `curl -s -f --insecure -H "api-token: $(get_admin_token)" https://localhost/api/v0/cfgmgmt/stats/checkin_counts_timeseries?days_ago=3 | jq`
1. Ensure you see around 5 node totals in all three buckets. You might see less than five because some of the nodes might have their date set in the future. :woman_shrugging: 
1. Update to this PR's changes with `rebuild components/config-mgmt-service && rebuild components/automate-gateway && rebuild components/ingest-service`
1. Adding some infra nodes with `chef_load_nodes 10`
1. Run this command to get the time series. `curl -s -f --insecure -H "api-token: $(get_admin_token)" https://localhost/api/v0/cfgmgmt/stats/checkin_counts_timeseries?days_ago=3 | jq`
1. Ensure you see around 10 total nodes for the latest date range. 

### :white_check_mark: Checklist
